harvard-lts/drs-import-management-service
# Message consumption error handling

**Message consumption error handling**
* * *

**GitHub Issue**: ([link](https://app.zenhub.com/workspaces/hdc-62163f60fb08c60011fa4795/issues/harvard-lts/hdc/192))

Other Relevant Links:
* [ActiveMQ docs - Message Redelivery and DLQ Handling](https://activemq.apache.org/message-redelivery-and-dlq-handling)
* [AmazonMQ docs - Elements and Their Attributes Permitted in Amazon MQ Configurations](https://docs.aws.amazon.com/amazon-mq/latest/developer-guide/permitted-attributes.html) - It is required for the correct operation of error handling that, as included in the activemq.xml configuration for the dockerized ActiveMQs for local dev environment, AmazonMQ includes the necessary configurations for DLQ policy and the redelivery plugin with fallback to DLQ.

# What does this Pull Request do?
It adds the management of errors in the consumption of messages in the application's listeners, managing through ACKs and NACKs (client-individual configuration) the successful or unsuccessful consumption of messages and sending to DLQ the unsuccessful ones.

A general refactor for listeners is also included, which extracts misplaced business logic from listeners to domain services.

Error handling in reading messages due to non-existent fields has also been reinforced, to avoid uncaught exceptions and send these messages to DLQ for further review.

# How should this be tested?

A description of what steps someone could take to:

* Make sure test .env variables are updated based on .env.example file
* Rebuild test environment: docker-compose -f docker-compose-test.yml build --no-cache
* Run the local test environment: `docker-compose -f docker-compose-test.yml up`
* Execute tests: `docker exec -it test_runner pytest test`

For a manual functional test:

* Make sure .env variables are updated based on .env.example file
* Rebuild dev environment: docker-compose -f docker-compose-dev.yml build --no-cache
* Run the local dev environment: `docker-compose -f docker-compose-dev.yml up`
* Send a malformed message to one of the queues (e.g. drs-ingest-status) :

![malformed_drs_ingest_status_message](https://user-images.githubusercontent.com/17010447/170450396-5dc57489-dfa7-4d22-9f31-ee8bbd91debc.png)

This should cause message consumption to fail due to missing required params.

* Wait until message is consumed and check that it is sent to DLQ:

![malformed_drs_ingest_status_queues](https://user-images.githubusercontent.com/17010447/170450743-3ea7545c-df44-48bf-9dc5-a364a6d9b8c6.png)

The message must be readable by accessing the DLQ:

![malformed_drs_ingest_status_dlq](https://user-images.githubusercontent.com/17010447/170450974-cfa8d7a7-a66e-4caa-8039-5be300f46526.png)

* dims.log file should contain entries like the following related with the unsuccessful message consumption:


```
Received frame: 'MESSAGE', len(body)=15

Received message from Process Queue. Message body: {'test': 'test'}. Message id: ID:9a426ce536c2-38897-1653553765765-8:1:1:1:1

Process Status message with id ID:9a426ce536c2-38897-1653553765765-8:1:1:1:1 does not contain 'package_id' field inside body

Setting message with id ID:9a426ce536c2-38897-1653553765765-8:1:1:1:1 as unacknowledged...

Sending frame: 'NACK'
```

# Test coverage
Yes/No: Are changes in this pull-request covered by:
- unit tests Yes
- integration tests Yes

# Interested parties
@jessjass @ives1227 @adaybujeda 

Keyword: resolved
